### ISSUE TITLE
Error in "npm install" step when building daedalus

### ISSUE THREAD

**Issue by singula** [input-output-hk/cardano-sl#2133]
I am trying to build the wallet in debian unstable.
I've followed the steps from
https://github.com/input-output-hk/cardano-sl/blob/master/docs/how-to/build-cardano-sl-and-daedalus-from-source-code.md
... and I got to the point where after running "./scripts/link-bridge.sh" I have to run "npm install".

That step however failed:

```
[nix-shell:~/daedalus]$ npm install
npm WARN package.json daedalus@0.8.0 No repository field.
npm WARN package.json daedalus@0.8.0 No license field.

> daedalus@0.8.0 preinstall /home/coincard/daedalus
> which electron || npm install electron@1.7.9

npm WARN package.json daedalus@0.8.0 No repository field.
npm WARN package.json daedalus@0.8.0 No license field.
npm WARN package.json daedalus-client-api@0.0.1 No repository field.
npm ERR! 404 Not Found
npm ERR! 404
npm ERR! 404 'types/node' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it
npm ERR! 404 It was specified as a dependency of 'electron'
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, or http url, or git url.

npm ERR! System Linux 4.13.0-1-amd64
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "install" "electron@1.7.9"
npm ERR! cwd /home/coincard/daedalus
npm ERR! node -v v6.12.0
npm ERR! npm -v 1.4.21
npm ERR! code E404
npm ERR!
npm ERR! Additional logging details can be found in:
npm ERR!     /home/coincard/daedalus/npm-debug.log
npm ERR! not ok code 0

npm ERR! daedalus@0.8.0 preinstall: `which electron || npm install electron@1.7.9`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the daedalus@0.8.0 preinstall script.
npm ERR! This is most likely a problem with the daedalus package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     which electron || npm install electron@1.7.9
npm ERR! You can get their info via:
npm ERR!     npm owner ls daedalus
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 4.13.0-1-amd64
npm ERR! command "/usr/bin/node" "/usr/bin/npm" "install"
npm ERR! cwd /home/coincard/daedalus
npm ERR! node -v v6.12.0
npm ERR! npm -v 1.4.21
npm ERR! code ELIFECYCLE
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian

npm ERR!
npm ERR! Additional logging details can be found in:
npm ERR!     /home/coincard/daedalus/npm-debug.log
npm ERR! not ok code 0
```

Documentation does not give any clue why this is happening and how to fix it.

**Comment by fschr** [comment]
I had a horrible time running with everything npm-related, as well. Most of the time, I was able to solve my issues by

1. Globally installing troublesome node modules *on my host* (i.e., not in nix shell) using `npm install -g some_module`
2. Copying the module directory from the global install location into daedalus/node_modules.

For example, the node module `scrypt` was not playing nicely. I fixed this issue by doing this

```
$ npm install -g scrypt
Installing scrypt....
....
Installed scrypt to /usr/lib/node_modules/scrypt
$ cp -r /usr/lib/node_modules/scrypt ~/daedalus/node_modules
```
I had to do this for a handful of modules. I also had to manually comment out broken asserts in the `fbjs-scripts` module. But alas, after much pain and effort, I was able to run `npm install` successfully.

**Comment by zmeel** [comment]
@fschr Did you also manage to connect the wallet to the back end?

**Comment by fschr** [comment]
@zmeel I did, but with these key differences from the guide

1. The guide doesn't mention this, but you need tmux in order to run `./scripts/launch/demo-with-wallet-api.sh` (this is the command which spins up a cluster).
2. After launching a cluster, I had to run `npm run dev` instead of `npm run hot-server` to actually get a wallet up and running.

However, my wallet got stuck on the "Syncing with blocks..." loading screen. I quickly realized this happened because my "/" partition was out of space, so the blocks couldn't finish syncing blocks. At this point, I said "f*ck it" and rm'd everything.

Here are my thoughts on the current state of Cardano-SL and Daedalus

* I should not have to build a web app in order to generate a private key and an address.
* I should not have to spin up a cluster, let alone "sync blocks," in order to generate a private key and an address.
* Fixing the previous two issues should be a number one priority at IOHK, imo. I understand that it is technically *possible* to generate a private key and address without building a web app or spinning up a cluster, but you have to dig through a sea of documentation to figure out how.
* Nix is a good way to get predictable builds, especially when relying on so many different pieces of software.
* However, the .nix files have a lot of issues. I think the developers miss these issues because they probably are all developing on the same platform. My guess is OS X.
* Nix takes way too much space for some people! I shouldn't have to download 5GB of libraries which I already have to build a simple web app. The current state of the documentation essentially reads *"If you don't use Nix, then good f\*cking luck!"*

**Comment by zmeel** [comment]
@fschr THanks! That's quite an eye opener. Building is terrible indeed. I'm trying to run it in a docker container on my Synology DS415+ with 8Gb RAM. I have plenty of disk space but speed is a problem. Building the whole thing takes more than 2 hours and changing only minor things and it all is going to build the whole project again. I'm struggling for days :-)

I have just run the `./scripts/launch/demo-with-wallet-api.sh` in a tmux session but then I get an error:
<pre>Invalid option `--vss-genesis'</pre>

Any ideas?

**Comment by fschr** [comment]
@zmeel RIP. I guess that means you have the wrong version of tmux or something. Can you post the full log including the command?

**Comment by zmeel** [comment]
@fschr I have to take another approach. Yesterday, I found out ([here](https://github.com/input-output-hk/cardano-sl/issues/2112)) you need xwindows to get the GIU to connect to the back end. As I'm running this in a Docker container on a Synology DS415+ I have don't have X. So, I'm going to take a different [approach](https://github.com/h4ck3rm1k3/daedalus/blob/master/README.md).

**Comment by zmeel** [comment]
Yes, I have seen that one and tried it but you still need xwindows for the front end wallet. For security reasons I'd rather build my own version. Although it is probably save, in theory he could have added some code to send my ADA to his wallet :-)
But I really do appreciate his effort...
I have the back end running in a Docker container on my Synology DS415+. I'm going to build the wallet in a Docker container on my MacBook Pro and use the GUI of the Mac to see the wallet in the container.

**Comment by morenoh149** [comment]
The error reads
```
> which electron || npm install electron@1.7.9

npm WARN package.json daedalus@0.8.0 No repository field.
npm WARN package.json daedalus@0.8.0 No license field.
npm WARN package.json daedalus-client-api@0.0.1 No repository field.
npm ERR! 404 Not Found
npm ERR! 404
npm ERR! 404 'types/node' is not in the npm registry.
npm ERR! 404 You should bug the author to publish it
npm ERR! 404 It was specified as a dependency of 'electron'
npm ERR! 404
npm ERR! 404 Note that you can also install from a
npm ERR! 404 tarball, folder, or http url, or git url.
```

Which looks like electron was not on the box so it tried to install it but that 404'd when it tried to install electron's dependency `types/node`.

Looking through their issue tracker suggests the problem is with the npm version being used here https://github.com/electron/electron/issues/10632 electron now requires at least npm v2.

@singula could you retry after updating the npm version?
try `npm install npm@latest -g` ([updating npm](https://docs.npmjs.com/getting-started/installing-node#2-update-npm))

https://cardanodocs.com/technical/wallet-frontend/#building-daedalus-client-api should instruct the user to install a specific version of npm.
